File: Array/medium/cpp/leaders_in_an_array.cpp
//
// Created by santosh on 12/20/25.
//

#include<bits/stdc++.h>
using namespace std;


class Solution
{
    public:
    vector<int> getLeadersInArray(vector<int>& nums)
    {
        vector<int> result;

        result.push_back(nums[nums.size()-1]);
        int lastElement=nums[nums.size()-1];
        int right=nums.size()-1;
        while (right>=0)
        {
            if (nums[right]>lastElement)
            {
                result.push_back(nums[right]);
                lastElement=nums[right];
            }
            right--;
        }

        reverse(result.begin(), result.end());

        return result;

    }
};
int main()
{
    Solution s;
    vector<int> nums={10, 22, 12, 3, 0, 6};
    vector<int> result = s.getLeadersInArray(nums);
    for (int i=0; i<result.size(); i++)
    {
         cout<<result[i]<<" ";
    }
}

File: Array/medium/python/leaders_in_an_array.py
# Leaders in an array
# An element is a leader if it is greater than all elements to its right.
# Created by santosh

class Solution:
    def get_leaders_in_array(self, nums: list[int]) -> list[int]:
        """
        Traverse from right to left keeping track of maximum.
        Time: O(n), Space: O(n)
        """
        result = []
        last_element = nums[-1]
        result.append(nums[-1])
        right = len(nums) - 2

        while right >= 0:
            if nums[right] > last_element:
                result.append(nums[right])
                last_element = nums[right]
            right -= 1

        result.reverse()
        return result


if __name__ == "__main__":
    s = Solution()
    nums = [10, 22, 12, 3, 0, 6]
    result = s.get_leaders_in_array(nums)
    print(" ".join(map(str, result)))
